### PR TITLE
[WIP] add uniswap bridge reserve

### PR DESCRIPTION
**Pull request by ducquangkstn**

### CODE REVIEW

File: contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+    uint public constant DEFAULT_FEE_BPS = 25;
+
+    address public kyberNetwork;
+
+    uint public feeBps = DEFAULT_FEE_BPS;
+
+    // Uniswap exchange contract for every listed token
+    // token -> exchange
+    mapping (address => address) public tokenExchange;
+
```
Comment (ilanDoron) [inline contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol:63]: extra line space

here and a few more places.

File: contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+            uint amountLessFee = deductFee(srcQty);
+            if (amountLessFee == 0) return (0, 0);
+
+
+            (uint reserveIn, uint reserveOut) = getReserves(uniswapPair, WETH, address(dest));
+
+            destQty = UniswapV2Library.getAmountOut(amountLessFee, reserveIn, reserveOut);
+        } else {
+            uniswapPair = IUniswapV2Pair(tokenExchange[address(src)]);
+            (uint reserveIn, uint reserveOut) = getReserves(uniswapPair, address(src), WETH);
```
Comment (ilanDoron) [inline contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol:366]: in and out are a bit confusing.
can it be named:
reserveSrcToken reserveDestToken
or its wrong?

File: contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+    {
+        onlyAdmin();
+        require(tokenExchange[address(token)] != address(0), "token is not listed");
+        address uniswapPair = tokenExchange[address(token)];
+
+        delete tokenExchange[address(token)];
+        
+        token.safeApprove(uniswapPair, 0);
+
+        TokenDelisted(token);
```
Comment (ilanDoron) [inline contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol:269]: can add pair address in this event.
and also index token

File: contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+        onlyAdmin();
+        require(bps <= 10000);
+
+        feeBps = bps;
+
+        FeeUpdated(bps);
+    }
+
+    event TokenListed(
+        IERC20 token,
```
Comment (ilanDoron) [inline contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol:236]: can index token

File: contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+        rate = calcRateFromQty(
+            srcQty, /* srcAmount */
+            destQty, /* destAmount */
+            getDecimals(src), /* srcDecimals */
+            getDecimals(dest) /* dstDecimals */
+        );
+    }
+
+    function getReserves(IUniswapV2Pair uniswapPair, address tokenA, address tokenB) internal view returns (uint reserveA, uint reserveB) {
+        (address token0,) = UniswapV2Library.sortTokens(tokenA, tokenB);
```
Comment (ilanDoron) [inline contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol:381]: can we have sorted tokens stored once per token listing.
And avoid extra call to sort token?

Comment (ilanDoron) [reply contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol:381]: but actually not sure its less gas. depends if sort token is external call or not.

File: contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+    }
+
+    /**
+     *   @dev called by kybernetwork to get settlement rate
+     */
+    function getConversionRate(
+        IERC20 src,
+        IERC20 dest,
+        uint256 srcQty,
+        uint256 /* blockNumber */
```
Comment (ilanDoron) [inline contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol:135]: nice. it works.
didn't know about it before.

File: contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+}
+
+interface IUniswapV2Pair {
+    function getReserves()
+        external
+        view
+        returns (
+            uint112 reserve0,
+            uint112 reserve1,
+            uint32 blockTimestampLast
```
Comment (ilanDoron) [inline contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol:50]: did u read what is this timstamp used for?

any chance we want to validate it?

File: contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+
+import "../../../IKyberReserve.sol";
+import "../../../IERC20.sol";
+import "../../../utils/WithdrawableNoModifiers.sol";
+import "../../../utils/Utils5.sol";
+import "../../../utils/zeppelin/SafeMath.sol";
+import "../../../utils/zeppelin/SafeERC20.sol";
+
+library UniswapV2Library {
+    using SafeMath for uint256;
```
Comment (ilanDoron) [inline contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol:11]: its best if we can have this library in a separate file. so we can easily update / upgrade if required.

File: contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+
+    function swap(
+        uint256 amount0Out,
+        uint256 amount1Out,
+        address to,
+        bytes calldata data
+    ) external;
+}
+
+interface IWETH {
```
Comment (ilanDoron) [inline contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol:61]: lets have this interface in a separate file

good chance we need it for other contracts later.
also if there is a "formal" interface its better.
from open zeppelin or somewhere else.

File: contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol [hunk 10]
```diff
@@ -0,0 +1,339 @@
+pragma solidity 0.6.6;
+
+import "../../../IKyberReserve.sol";
```
Comment (ilanDoron) [inline contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol:3]: maybe we can  have bridgeReserves as a separate folder. not inside reserves.
Seems to like we can have reserves as Kybers reserves (fpr, apr etc)
and bridgeReserves is a different thing.

but not sure.
lets discuss in next SC meeting.

File: contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol (hunk 5 above)
Comment (ducquangkstn) [reply contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol]: This is internal library call. So it does not cost much

**Comment by ducquangkstn** [comment]
Change to using best paths by querying rate. Closing

**Pull request closed without merging**